tower-http/src/classify.rs

```diff
@@ -0,0 +1,313 @@
+    type FailureClass = T;
+
+    fn classify_eos(self, _trailers: Option<&HeaderMap>) -> Result<(), Self::FailureClass> {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+}
+
+/// The default classifier used for normal HTTP responses.
+///
+/// Responses with a `5xx` status code are failures, all others as successes.
+#[derive(Clone, Debug, Default)]
+pub struct ServerErrorsAsFailures {
+    _priv: (),
+}
+
+impl ServerErrorsAsFailures {
+    /// Create a new [`ServerErrorsAsFailures`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a [`MakeClassifier`] that produces `ServerErrorsAsFailures`.
+    ///
+    /// This is a convenience function that simply calls `SharedClassifier::new`.
+    pub fn make_classify<E>() -> SharedClassifier<Self> {
+        SharedClassifier::new::<E>(Self::new())
+    }
+}
+
+impl<E> ClassifyResponse<E> for ServerErrorsAsFailures {
+    type FailureClass = StatusCode;
+    type ClassifyEos = NeverClassifyEos<StatusCode>;
+
+    fn classify_response<B>(
+        self,
+        res: &Response<B>,
+    ) -> ClassifiedResponse<Self::FailureClass, Self::ClassifyEos>
+    where
+        B: Body,
+    {
+        if res.status().is_server_error() {
+            ClassifiedResponse::Ready(Err(res.status()))
+        } else {
+            ClassifiedResponse::Ready(Ok(()))
+        }
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        StatusCode::INTERNAL_SERVER_ERROR
+    }
+}
+
+/// Response classifier for gRPC responses.
+///
+/// gRPC doesn't use normal HTTP statuses for indicating success or failure but instead a special
+/// header that might appear in a trailer.
+///
+/// Responses where `grpc-status` is 0 are successes, all others are failures.
+#[derive(Debug, Clone, Default)]
+pub struct GrpcErrorsAsFailures {
+    _priv: (),
+}
+
+impl GrpcErrorsAsFailures {
+    /// Create a new [`GrpcErrorsAsFailures`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a [`MakeClassifier`] that produces `GrpcErrorsAsFailures`.
+    ///
+    /// This is a convenience function that simply calls `SharedClassifier::new`.
+    pub fn make_classify<E>() -> SharedClassifier<Self> {
+        SharedClassifier::new::<E>(Self::new())
+    }
+}
+
+impl<E> ClassifyResponse<E> for GrpcErrorsAsFailures {
+    type FailureClass = i32;
+    type ClassifyEos = GrpcEosErrorsAsFailures;
+
+    fn classify_response<B>(
+        self,
+        res: &Response<B>,
+    ) -> ClassifiedResponse<Self::FailureClass, Self::ClassifyEos>
+    where
+        B: Body,
+    {
+        if let Some(classification) = classify_grpc_metadata(res.headers()) {
+            ClassifiedResponse::Ready(classification)
+        } else {
+            ClassifiedResponse::RequiresEos(GrpcEosErrorsAsFailures { _priv: () })
+        }
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        13
```

Comment:
I'm not familiar with grpc so much, what does 13 refer to?

Reply:
Its gRPC's version of "internal server error". I added a link that should explain it.

Reply:
Thanks!